### DISCUSSION TITLE
General development topic

### DISCUSSION THREAD

**Discussion by jamieyello** [MNES-Emulator/MNES#2] [General]
@Morbidous I'll work on adding the next couple of CPU methods. I'm around 70/151 opcodes implemented, lmk and I'll show you my workflow. I can explain how the code works as well.

I expect we'll gain some public support after some fun experimental features, but for now my main goal is to boot SMB1.

**Comment by jamieyello** [comment]
[This](https://www.masswerk.at/6502/6502_instruction_set.html) has been my goto sheet for adding instructions.

[This](https://github.com/MNES-Emulator/MNES/blob/master/MNES.Tests/Machine/MachineTests.cs) has been the main test I've been using. All I do is right click on it and debug and it tells me what I need to do next. It's also just a good entry point to understand how things are working.

[This](https://github.com/MNES-Emulator/MNES/blob/master/MNES.Tests/Resources/Test%20Roms/other/nestest_log.txt) is the proper output from another emulator that fully supports all CPU instructions. By comparing this output with the debug output, we can find out if anything is not working correctly, and the goal is to make it match 1:1. I just keep this open off to the side while I work.

**Comment by jamieyello** [comment]
I'm also going to commit to making a lot of smaller commits after I do something instead of one big one, who cares about commit spam

**Comment by jamieyello** [comment]
I added a logo gif ~~**_but it might be too flashy_**~~
https://github.com/MNES-Emulator/MNES

**Comment by jamieyello** [comment]
[106/151 opcodes](https://github.com/MNES-Emulator/MNES/commit/5ef3f3330440eb02413575ce281d24cb89009463) implemented

The remaining ones should be easy since they're variations of previous ones.

**Comment by jamieyello** [comment]
@Morbidous The CPU is fully implemented. There are still some bugs to sort out, and I'll do that quickly tomorrow, but now is a good time to look into building out the APU, the PPU, or IO registers from scratch if you're interested in any of those things. I know as much about them as you do, so I won't have such a head start.

We can also start looking into the other tests roms provided to see how they can help and set up proper unit tests for them. I also have things set up to the point where we can run games that use [mapper 0](https://www.nesdev.org/wiki/Mapper) like Donkey Kong and SMB1.

I can slow down and start working on other things like Godot UI.

**Comment by Morbidous** [comment]
I can only do so much with limited time and energy so it's no problem if I just do random small things for now. I would need to grok the system a bit better to do any of my own implementation (and I especially don't want to trample all over your work either). I'll have more time on the weekend too maybe (depending what I use that time for though).

**Comment by jamieyello** [comment]
All PPU registers are implemented, now I just need to implement scanline rendering. I've thought about how I would do it with Godot tilemaps, and it would be possible, it would just be easier for now to do it through software rendering pixel by pixel (which is more or less how the NES did it anyway)

Even though you can easily derive a BG tilemap and sprite layout from the NES PPU VRAM/OAM and draw a _pretty close_ rendering, the CPU can change anything after a scanline, which complicates things. It can swap tilesets, palettes, change the X offset of the BG, and things like that.

**Comment by jamieyello** [comment]
Got some fancy animation that fades the logo in from off-screen, I used the built-in Godot file manager (even though I hate those, I prefer the OS file manager)

The alignment is all wrong but I've tried to fix it for too long.

I want to do some stupidly polished UI (for emulator standards) with an overworld to run around in where you select your game. I want to have it so once you beat a game, you unlock that character from that game to play with in other games, and can also use that character in the overworld. I want to create a mod for each supported game that will hook into wall collision and work with other games, including a "green screen" world that lets the emulator get character graphics. That would be crazy.

![](https://raw.githubusercontent.com/MNES-Emulator/MNES/f1777864e20d115521be7529f3a182acfbd6534f/Resources/Images/Previews/MNES%20UI%201.gif)

**Comment by jamieyello** [comment]
@Morbidous I added instructions on how to launch the GUI [here](https://github.com/MNES-Emulator/MNES/blob/master/CONTRIBUTING.md) if you're interested.

**Comment by jamieyello** [comment]
With the most recent PPU work, the game actually runs and attempts to draw things normally. There are just a lot of bugs preventing it from showing anything correctly. It will be a bit more difficult debugging the PPU, since projects like Nintendulator don't expose what's going on in the PPU as much as they do the CPU, but I will slowly try to work through it.
